src/external_trait_impls/serde.rs

```diff
@@ -0,0 +1,116 @@
+use core::fmt;
+use core::marker::PhantomData;
+
+use serde_::de::{Deserialize, Deserializer, SeqAccess, Visitor};
+use serde_::ser::{Serialize, Serializer};
+
+use crate::Vc;
+
+impl<T> Serialize for Vc<T>
+where
+    T: Serialize,
+{
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.collect_seq(self)
+    }
+}
+
+#[inline]
+fn cautious_size_hint(hint: Option<usize>) -> usize {
+    core::cmp::min(hint.unwrap_or(0), 4096)
+}
+
+impl<'de, T> Deserialize<'de> for Vc<T>
+where
+    T: Deserialize<'de>,
+{
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        struct VcVisitor<T> {
+            marker: PhantomData<T>,
+        }
+        impl<'de, T> Visitor<'de> for VcVisitor<T>
+        where
+            T: Deserialize<'de>,
+        {
+            type Value = Vc<T>;
+            fn expecting(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+                formatter.write_str("a sequence")
+            }
+            fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+            where
+                A: SeqAccess<'de>,
+            {
+                let mut values = Vc::with_capacity(cautious_size_hint(seq.size_hint()));
+                while let Some(value) = seq.next_element()? {
+                    values.push(value);
+                }
+                Ok(values)
+            }
+        }
+        let visitor = VcVisitor {
+            marker: PhantomData,
+        };
+        deserializer.deserialize_seq(visitor)
+    }
+}
```

Comment:
The main `serde` implementation of `Deserialize` for `Vec<T>` also includes `deserialize_in_place` here:
https://github.com/serde-rs/serde/blob/24e6acbfaeb18af978012b904209632f012eb54d/serde/src/de/impls.rs#L883

Might be good to include that here too?

Reply:
I'm on it! And, I have questions. 😄

In their implementation, it uses the `reserve` method.

```
                let hint = size_hint::cautious(seq.size_hint());
                if let Some(additional) = hint.checked_sub(self.0.len()) {
                    self.0.reserve(additional);
                }
```

Which is understandable. However, `Vc`'s `reserve` does the `carry_all` in case there is a resizing, which is not a good showcase for `Vc`😅.

The questions are
1. Should we use `reserve`? Or just let the `Vc`'s `push` magic handle the work?
2. How do I test `deserialize_in_place`?

Reply:
I think it's fine to use `reserve` here. The carry cost should be small here compared to the cost of the deserialization.

Reply:
As for how to test it, I'm not entirely sure... What do they do?